### PR TITLE
Dynamically change communication protocol

### PR DESCRIPTION
**Pull request by imurchie**
There have been problems where the wrong communication system is used. This PR changes the logic so that the initial condition is based on what seems to be right, but then if the error comes in that the domain is not know, we switch to the other protocol.

See: 
https://github.com/appium/appium/issues/12518
https://github.com/appium/appium/issues/12510

### CODE REVIEW

File: lib/rpc-client.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import events from 'events';
+import log from './logger';
+import RemoteMessages from './remote-messages';
+
+
+export default class RpcClient extends events.EventEmitter {
+  constructor () {
```
Comment (mykola-mokhnach) [inline lib/rpc-client.js]: this is probably redundant

File: lib/rpc-client.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    if (!this.remoteMessages) {
+      this.remoteMessages = new RemoteMessages(isTargetBased);
+    } else {
+      this.remoteMessages.setCommunicationProtocol(isTargetBased);
+    }
+  }
+
+  async send (command, opts = {}) {
+    try {
+      return await this.sendMessage(command, opts);
```
Comment (mykola-mokhnach) [inline lib/rpc-client.js]: should we declare an empty sendMessage method in this class to show that it must be overridden by subclasses?

File: lib/remote-debugger-rpc-client.js [hunk 3]
```diff
@@ -41,10 +43,18 @@ export default class RemoteDebuggerRpcClient {
...
 
-    const targetBased = isTargetBased(isSafari, platformVersion);
-    this.remoteMessages = new RemoteMessages(targetBased);
-    this.messageHandler = new RpcMessageHandler(this.specialMessageHandlers, targetBased);
-    log.debug(`Using ${targetBased ? 'Target-based' : 'full Web Inspector protocol'} communication`);
+    // start with a best guess for the protocol
+    this.setCommunicationProtocol(isTargetBased(isSafari, platformVersion));
+  }
+
+  setCommunicationProtocol (isTargetBased = false) {
```
Comment (mykola-mokhnach) [inline lib/remote-debugger-rpc-client.js]: how about having setter/getter instead?

Comment (imurchie) [reply lib/remote-debugger-rpc-client.js]: I'm not a big fan of getters/setters that do other things.

File: lib/rpc-client.js (hunk 1 above)
Comment (imurchie) [reply lib/rpc-client.js]: It is if EventEmitter is removed (and I'm not sure why it is an EventEmitter).

File: lib/rpc-client.js (hunk 2 above)
Comment (imurchie) [reply lib/rpc-client.js]: Done!

**Pull request merged**